reconcile/openshift_base.py

```diff
@@ -568,6 +647,177 @@ def _realize_resource_data(
+                    continue
+
+                logging.debug("CURRENT: " + OR.serialize(OR.canonicalize(c_item.body)))
+        else:
+            logging.debug("CURRENT: None")
+
+        logging.debug("DESIRED: " + OR.serialize(OR.canonicalize(d_item.body)))
+
+        try:
+            privileged = data["use_admin_token"].get(name, False)
+            apply(
+                dry_run,
+                oc_map,
+                cluster,
+                namespace,
+                resource_type,
+                d_item,
+                wait_for_namespace,
+                recycle_pods,
+                privileged,
+            )
+            action = {
+                "action": ACTION_APPLIED,
+                "cluster": cluster,
+                "namespace": namespace,
+                "kind": resource_type,
+                "name": d_item.name,
+                "privileged": privileged,
+            }
+            actions.append(action)
+        except StatusCodeError as e:
+            ri.register_error()
+            err = (
+                str(e)
+                if resource_type != "Secret"
+                else f"error applying Secret {d_item.name}: REDACTED"
+            )
+            msg = (
+                f"[{cluster}/{namespace}] {err} "
+                + f"(error details: {d_item.error_details})"
+            )
+            logging.error(msg)
+
+    # current items
+    for name, c_item in data["current"].items():
+        d_item = data["desired"].get(name)
+        if d_item is not None:
+            continue
+
+        if c_item.has_qontract_annotations():
+            if caller and c_item.caller != caller:
+                continue
+        elif not take_over:
+            # this is reached when the current resources:
+            # - does not have qontract annotations (not managed)
+            # - not taking over all resources of the current kind
+            msg = (
+                f"[{cluster}/{namespace}] skipping " + f"{resource_type}/{c_item.name}"
+            )
+            logging.debug(msg)
+            continue
+
+        if c_item.has_owner_reference():
+            continue
+
+        try:
+            privileged = data["use_admin_token"].get(name, False)
+            delete(
+                dry_run,
+                oc_map,
+                cluster,
+                namespace,
+                resource_type,
+                name,
+                enable_deletion,
+                privileged,
+            )
+            action = {
+                "action": ACTION_DELETED,
+                "cluster": cluster,
+                "namespace": namespace,
+                "kind": resource_type,
+                "name": name,
+                "privileged": privileged,
+            }
+            actions.append(action)
+        except StatusCodeError as e:
+            ri.register_error()
+            msg = "[{}/{}] {}".format(cluster, namespace, str(e))
+            logging.error(msg)
```

Comment:
Now that we are creating a new method, can we make the logic flow less deep nested?
There is a generic differ can be used to diff current and desired dict, https://github.com/app-sre/qontract-reconcile/blob/2fa08e3f01b2d0aabde8ce8bf3247b41d6d320ef/reconcile/utils/differ.py#L41C5-L41C18.

```python
    diff_result = differ.diff_mappings(
        data["currrent"], data["desired"], equal=three_way_merge_patch_diff_using_hash
    )
```

Then can process each part:

```python
    diff_result.add
    diff_result.delete
    diff_result.change
    diff_result.identical
```

It will be good we don't have any `continue` in this method to make it a flat flow.

Reply:
I wasn't aware of this differ class. I refactored the method to make use of it and added tests as well. Thanks for doing it and letting me know. PTAL.

Reply:
Nice!